infrastructure/src/main/java/org/corfudb/infrastructure/health/HealthStatus.java

```diff
@@ -0,0 +1,127 @@
+            initStatus = InitStatus.NOT_INITIALIZED;
+            runtimeHealthIssues.clear();
+        }
+        initHealthIssues.add(issue);
+    }
+
+    /**
+     * Remove the init health issue -- report to health monitor that the component is initialized.
+     * @param issue An init issue
+     */
+    public void resolveInitHealthIssue(Issue issue) {
+        verifyComponent(issue);
+        if (initStatus == InitStatus.NOT_INITIALIZED) {
+            initStatus = InitStatus.INITIALIZED;
+        }
+        initHealthIssues.remove(issue);
+    }
+
+    /**
+     * Add runtime health issue -- health monitor now expects the component's runtime issue to recover.
+     * If component has been not been previously initialized (INITIALIZED), adding the runtime issue will
+     * result in exception. If the runtime issue is already present in the HealthStatus, this is a
+     * NOOP and the previously added issue maintains it's position in the runtimeHealthIssues
+     * LinkedHashSet.
+     * @param issue An init issue
+     */
+    public void addRuntimeHealthIssue(Issue issue) {
+        verifyComponent(issue);
+        if (initStatus != InitStatus.INITIALIZED) {
+            throw new IllegalStateException("Runtime health issue can only be reported if the component is initialized");
+        }
+        runtimeHealthIssues.add(issue);
+    }
+
+    /**
+     * Remove the runtime health issue -- report to health monitor that this issue is resolved.
+     * @param issue An init issue
+     */
+    public void resolveRuntimeHealthIssue(Issue issue) {
+        verifyComponent(issue);
+        runtimeHealthIssues.remove(issue);
+    }
+
+    public boolean isInitHealthy() {
+        return initHealthIssues.isEmpty() && initStatus == InitStatus.INITIALIZED;
+    }
+
+    public boolean isRuntimeHealthy() {
+        return runtimeHealthIssues.isEmpty() && initStatus == InitStatus.INITIALIZED;
+    }
+
+    private void verifyComponent(Issue issue) {
+        if (issue.getComponent() != this.component) {
+            String msg = String.format("Only the issues belonging to %s are allowed but got: %s", this.component,
+                    issue.getComponent());
+
+            throw new IllegalStateException(msg);
+        }
+    }
+    /**
+     * Get the last reported unique runtime issue (If the runtime issue is already present
+     * in the HealthStatus and added again, it is not considered the latest since it's not resolved
+     * yet).
+     * @return A possible runtime health issue
+     */
+    public Optional<Issue> getLatestRuntimeIssue() {
+        if (runtimeHealthIssues.isEmpty()) {
+            return Optional.empty();
+        }
+        return runtimeHealthIssues.stream().skip(runtimeHealthIssues.size() - 1).findFirst();
+    }
+
+    public boolean containsIssue(Issue issue) {
+        return initHealthIssues.contains(issue) || runtimeHealthIssues.contains(issue);
+    }
+
+    enum InitStatus {
```

Comment:
dont we need a similar enum for RuntimeStatus?

Reply:
Runtime status existence solely depends on the status of init, so it does not really need any states